From 026d5c73780c5130062eb3b4015e9605080dcddf Mon Sep 17 00:00:00 2001
From: Wenyou Yang <wenyou.yang@atmel.com>
Date: Fri, 30 Jan 2015 11:46:48 +0800
Subject: [PATCH] driver: act8865: move the act8865 workaround code to the
 driver file

To decrease the the duplicated code, move the act8865 workaround
from the board file to the driver file act8865.c.

The board implements int at91_board_act8865_set_reg_voltage(void) interface.

Add a new config to select disabling the ACT8865's I2C.

Signed-off-by: Wenyou Yang <wenyou.yang@atmel.com>
---
 Config.in                                 | 18 ++++++++++------
 board/sama5d3_xplained/sama5d3_xplained.c | 19 ++--------------
 board/sama5d4_xplained/sama5d4_xplained.c | 19 ++--------------
 board/sama5d4ek/sama5d4ek.c               | 19 ++--------------
 driver/act8865.c                          | 36 +++++++++++++++++++++++--------
 driver/driver_cpp.mk                      |  4 ++--
 include/act8865.h                         |  3 ++-
 include/board.h                           |  2 ++
 main.c                                    |  5 +++++
 9 files changed, 56 insertions(+), 69 deletions(-)

diff --git a/Config.in b/Config.in
index 790135d..5780124 100644
--- a/Config.in
+++ b/Config.in
@@ -487,18 +487,26 @@ endmenu
 menu "Board's Workaround Options"
 	depends on CONFIG_HAS_PMIC_ACT8865
 
-config CONFIG_ACT8865
+config CONFIG_DISABLE_ACT8865_I2C
 	bool "Disable the ACT8865 I2C interface"
 	depends on CONFIG_HAS_PMIC_ACT8865
-	select CONFIG_TWI
-	default n
+	select CONFIG_ACT8865
+	default y
 	help
 	  This interface let you to disable the ACT8865's I2C interface.
 	  There are some issue about active-semi ACT8865's I2C interface,
 	  which affects the other chip which stands on the same I2C bus, such as MXT touchscreen.
 
+endmenu
+
+config CONFIG_ACT8865
+	bool "PMIC (ACT8865) Support"
+	depends on CONFIG_HAS_PMIC_ACT8865
+	select CONFIG_TWI
+	default n
+
 choice
-	prompt "PMIC on TWI Select"
+	prompt "PMIC (ACT8865) on TWI Select"
 	depends on CONFIG_ACT8865
 	default CONFIG_PMIC_ON_TWI0
 	help
@@ -526,8 +534,6 @@ config	CONFIG_PMIC_ON_TWI3
 
 endchoice
 
-endmenu
-
 menu "Board Hardware Information Options"
 	depends on CONFIG_HAS_HW_INFO
 
diff --git a/board/sama5d3_xplained/sama5d3_xplained.c b/board/sama5d3_xplained/sama5d3_xplained.c
index 778cdf8..fbaf35e 100644
--- a/board/sama5d3_xplained/sama5d3_xplained.c
+++ b/board/sama5d3_xplained/sama5d3_xplained.c
@@ -44,7 +44,6 @@
 #include "arch/at91_pio.h"
 #include "arch/at91_ddrsdrc.h"
 #include "sama5d3_xplained.h"
-#include "twi.h"
 #include "act8865.h"
 
 static void at91_dbgu_hw_init(void)
@@ -319,8 +318,8 @@ unsigned int at91_twi2_hw_init(void)
 }
 #endif
 
-#ifdef CONFIG_ACT8865
-static int sama5d3ek_act8865_set_reg_voltage(void)
+#if defined(CONFIG_DISABLE_ACT8865_I2C)
+int at91_board_act8865_set_reg_voltage(void)
 {
 	unsigned char reg, value;
 	int ret;
@@ -413,20 +412,6 @@ void hw_init(void)
 	/* Initialize MPDDR Controller */
 	ddramc_init();
 #endif
-
-#ifdef CONFIG_TWI
-	twi_init();
-#endif
-
-#ifdef CONFIG_ACT8865
-	/* Set ACT8865 output voltage */
-	sama5d3ek_act8865_set_reg_voltage();
-
-	/* Dsiable ACT8865 I2C interface */
-	if (act8865_workaround_disable_i2c())
-		while (1)
-			;
-#endif
 }
 #endif /* #ifdef CONFIG_HW_INIT */
 
diff --git a/board/sama5d4_xplained/sama5d4_xplained.c b/board/sama5d4_xplained/sama5d4_xplained.c
index 6405fa3..5b3cb5c 100644
--- a/board/sama5d4_xplained/sama5d4_xplained.c
+++ b/board/sama5d4_xplained/sama5d4_xplained.c
@@ -47,7 +47,6 @@
 #include "sama5d4_xplained.h"
 #include "tz_utils.h"
 #include "matrix.h"
-#include "twi.h"
 #include "act8865.h"
 
 #if defined(CONFIG_REDIRECT_ALL_INTS_AIC)
@@ -510,8 +509,8 @@ unsigned int at91_twi3_hw_init(void)
 }
 #endif
 
-#ifdef CONFIG_ACT8865
-static int sama5d4ek_act8865_set_reg_voltage(void)
+#if defined(CONFIG_DISABLE_ACT8865_I2C)
+int at91_board_act8865_set_reg_voltage(void)
 {
 	unsigned char reg, value;
 	int ret;
@@ -659,20 +658,6 @@ void hw_init(void)
 	/* Reset HDMI SiI9022 */
 	SiI9022_hw_reset();
 
-#ifdef CONFIG_TWI
-	twi_init();
-#endif
-
-#ifdef CONFIG_ACT8865
-	/* Set ACT8865 output voltage */
-	sama5d4ek_act8865_set_reg_voltage();
-
-	/* Dsiable ACT8865 I2C interface */
-	if (act8865_workaround_disable_i2c())
-		while (1)
-			;
-#endif
-
 #ifdef CONFIG_USER_HW_INIT
 	hw_init_hook();
 #endif
diff --git a/board/sama5d4ek/sama5d4ek.c b/board/sama5d4ek/sama5d4ek.c
index af32346..40e627c 100644
--- a/board/sama5d4ek/sama5d4ek.c
+++ b/board/sama5d4ek/sama5d4ek.c
@@ -48,7 +48,6 @@
 #include "sama5d4ek.h"
 #include "tz_utils.h"
 #include "matrix.h"
-#include "twi.h"
 #include "act8865.h"
 
 #if defined(CONFIG_REDIRECT_ALL_INTS_AIC)
@@ -522,8 +521,8 @@ unsigned int at91_twi3_hw_init(void)
 }
 #endif
 
-#ifdef CONFIG_ACT8865
-static int sama5d4ek_act8865_set_reg_voltage(void)
+#if defined(CONFIG_DISABLE_ACT8865_I2C)
+int at91_board_act8865_set_reg_voltage(void)
 {
 	unsigned char reg, value;
 	int ret;
@@ -697,20 +696,6 @@ void hw_init(void)
 	/* Reset HDMI SiI9022 */
 	SiI9022_hw_reset();
 
-#ifdef CONFIG_TWI
-	twi_init();
-#endif
-
-#ifdef CONFIG_ACT8865
-	/* Set ACT8865 output voltage */
-	sama5d4ek_act8865_set_reg_voltage();
-
-	/* Dsiable ACT8865 I2C interface */
-	if (act8865_workaround_disable_i2c())
-		while (1)
-			;
-#endif
-
 #ifdef CONFIG_USER_HW_INIT
 	hw_init_hook();
 #endif
diff --git a/driver/act8865.c b/driver/act8865.c
index d567578..f51a8a7 100644
--- a/driver/act8865.c
+++ b/driver/act8865.c
@@ -27,6 +27,7 @@
  */
 #include "common.h"
 #include "hardware.h"
+#include "board.h"
 #include "twi.h"
 #include "act8865.h"
 #include "debug.h"
@@ -132,6 +133,13 @@ int act8865_set_reg_voltage(unsigned char volt_reg, unsigned char value)
 	return 0;
 }
 
+int act8865_check_i2c_disabled(void)
+{
+	unsigned char data = 0;
+
+	return act8865_read(SYS_0, &data);
+}
+
 #ifdef CONFIG_PM_PMIC
 static int act8865_set_reg_mode(unsigned char mode_reg, unsigned mode)
 {
@@ -179,6 +187,7 @@ int act8865_set_power_saving_mode(void)
 
 /*--------------------- ACT8865 Workaround -----------------------*/
 
+#if defined(CONFIG_DISABLE_ACT8865_I2C)
 /*
  * Special Registers
  */
@@ -224,17 +233,9 @@ static int act8865_disable_i2c_sequence(unsigned char data)
 		return -1;
 
 	return 0;
-
 }
 
-int act8865_check_i2c_disabled(void)
-{
-	unsigned char data = 0;
-
-	return act8865_read(SYS_0, &data);
-}
-
-int act8865_workaround_disable_i2c(void)
+static int act8865_workaround_disable_i2c(void)
 {
 	unsigned char value;
 	unsigned char i;
@@ -260,3 +261,20 @@ int act8865_workaround_disable_i2c(void)
 
 	return 0;
 }
+
+void act8865_workaround(void)
+{
+	if (!twi_init_done)
+		twi_init();
+
+	/* Set ACT8865 REG output voltage */
+	at91_board_act8865_set_reg_voltage();
+
+	/* Disable ACT8865 I2C interface, if failed, don't go on */
+	if (act8865_workaround_disable_i2c()) {
+		dbg_info("ACT8865: Failed to disable I2C interface\n");
+		while (1)
+			;
+	}
+}
+#endif
diff --git a/driver/driver_cpp.mk b/driver/driver_cpp.mk
index 54b6022..7cbbe79 100644
--- a/driver/driver_cpp.mk
+++ b/driver/driver_cpp.mk
@@ -217,8 +217,8 @@ ifeq ($(CONFIG_TWI3), y)
 CPPFLAGS += -DCONFIG_TWI3
 endif
 
-ifeq ($(CONFIG_ACT8865), y)
-CPPFLAGS += -DCONFIG_ACT8865
+ifeq ($(CONFIG_DISABLE_ACT8865_I2C), y)
+CPPFLAGS += -DCONFIG_DISABLE_ACT8865_I2C
 endif
 
 ifeq ($(CONFIG_PM), y)
diff --git a/include/act8865.h b/include/act8865.h
index f59d91c..2c55aab 100644
--- a/include/act8865.h
+++ b/include/act8865.h
@@ -88,7 +88,8 @@
  */
 extern int act8865_set_reg_voltage(unsigned char volt_reg, unsigned char value);
 extern int act8865_check_i2c_disabled(void);
-extern int act8865_workaround_disable_i2c(void);
+
 extern int act8865_set_power_saving_mode(void);
+extern void act8865_workaround(void);
 
 #endif
diff --git a/include/board.h b/include/board.h
index c0362a3..52698e0 100644
--- a/include/board.h
+++ b/include/board.h
@@ -109,4 +109,6 @@ extern void at91_disable_mac_clock(void);
 
 extern void at91_disable_smd_clock(void);
 
+extern int at91_board_act8865_set_reg_voltage(void);
+
 #endif /* #ifndef __BOARD_H__ */
diff --git a/main.c b/main.c
index fbc9837..d87b32b 100644
--- a/main.c
+++ b/main.c
@@ -39,6 +39,7 @@
 #include "board_hw_info.h"
 #include "tz_utils.h"
 #include "pm.h"
+#include "act8865.h"
 
 extern int load_kernel(struct image_info *img_info);
 
@@ -157,6 +158,10 @@ int main(void)
 	at91_board_pm();
 #endif
 
+#ifdef CONFIG_DISABLE_ACT8865_I2C
+	act8865_workaround();
+#endif
+
 	init_loadfunction();
 
 	ret = (*load_image)(&image);
-- 
1.9.1

